services/slam/fake/slam.go

```diff
@@ -0,0 +1,160 @@
+	pngTemplate           = "slam/example_cartographer_outputs/image_map/image_map_%d.png"
+	positionTemplate      = "slam/example_cartographer_outputs/position/position_%d.txt"
+)
+
+func init() {
+	registry.RegisterService(
+		slam.Subtype,
+		model,
+		registry.Service{
+			Constructor: func(
+				ctx context.Context,
+				_ registry.Dependencies,
+				config config.Service,
+				logger golog.Logger,
+			) (interface{}, error) {
+				return &SLAM{Name: config.Name, dataCount: 1}, nil
+			},
+		},
+	)
+}
+
+var _ = slam.Service(&SLAM{})
+
+// SLAM is a fake slam that returns generic data.
+type SLAM struct {
+	generic.Echo
+	Name      string
+	dataCount int
+}
+
+func (slamSvc *SLAM) GetMap(ctx context.Context, name, mimeType string, cp *referenceframe.PoseInFrame,
+	include bool, extra map[string]interface{},
+) (string, image.Image, *vision.Object, error) {
+	var err error
+	var img image.Image
+	var vObj *vision.Object
+
+	// Increment data after getMap call
+	slamSvc.incrementDataCount()
+
+	switch mimeType {
+	case rdkutils.MimeTypePCD:
+		f, err := os.Open(artifact.MustPath(fmt.Sprintf(pcdTemplate, slamSvc.dataCount)))
+		if err != nil {
+			return "", nil, nil, err
+		}
+		defer utils.UncheckedErrorFunc(f.Close)
+		pc, err := pointcloud.ReadPCDToBasicOctree(f)
+		if err != nil {
+			return "", nil, nil, err
+		}
+		vObj, err = vision.NewObject(pc)
+		if err != nil {
+			return "", nil, nil, err
+		}
+
+	case rdkutils.MimeTypeJPEG:
+		img, err = rimage.NewImageFromFile(artifact.MustPath(fmt.Sprintf(pngTemplate, slamSvc.dataCount)))
+
+	default:
+		return "", nil, nil, errors.New("received invalid mimeType for GetMap call")
+	}
+
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	return mimeType, img, vObj, nil
+}
+
+func (slamSvc *SLAM) Position(ctx context.Context, name string, extra map[string]interface{}) (*referenceframe.PoseInFrame, error) {
+	data, err := os.ReadFile(artifact.MustPath(fmt.Sprintf(positionTemplate, slamSvc.dataCount)))
+	if err != nil {
+		return nil, err
+	}
+
+	substrings := strings.Split(string(data), " | ")
+
+	point, err := extract(strings.Split(substrings[0], " "))
+	if err != nil {
+		return nil, err
+	}
+	xyz := r3.Vector{X: point[0], Y: point[1], Z: point[2]}
+
+	orientations, err := extract(strings.Split(substrings[1], " "))
+	if err != nil {
+		return nil, err
+	}
+
+	ori := spatialmath.NewR4AA()
+	ori.RX = orientations[0]
+	ori.RY = orientations[1]
+	ori.RZ = orientations[2]
+	ori.Theta = orientations[3]
+	pose := spatialmath.NewPose(xyz, ori)
+
+	pInFrame := referenceframe.NewPoseInFrame(name, pose)
+
+	return pInFrame, nil
+}
+
+func extract(strings []string) ([]float64, error) {
+	elems := make([]float64, len(strings))
+	for i, v := range strings {
+		x, err := strconv.ParseFloat(v, 64)
+		if err != nil {
+			return nil, err
+		}
+		elems[i] = x
+	}
+	return elems, nil
+}
+
+func (slamSvc *SLAM) GetInternalState(ctx context.Context, name string) ([]byte, error) {
```

Comment:
We probably could leave this unimplemented, since it's not used by the UI, but fine to leave as is, in case we need it in the future.

Reply:
Yep, I wouldn't have implemented it but Jeremy already did the work so I figured it didn't hurt to leave it in, even if the UI won't use it.